### ISSUE TITLE
define data format

### ISSUE THREAD

**Issue by tpreusse** [nzzdev/red-zh-vote-crawler#1]
Currently it simply fetches html tables and converts them to javascript objects.

Please provide feedback and suggestions for the data structure.
### Example

[![screen shot 2015-03-31 at 16 53 13](https://cloud.githubusercontent.com/assets/410211/6921687/b39db4c0-d7c6-11e4-8eae-115f5d6eae9d.png)](http://www.wahlen.zh.ch/wahlen/kr2011_medieninfo/viewer.php?menu=listen_kanton)
Becomes:

``` js
[ { 'Listen Nr.': '1',
    'Listen Bez.': 'SVP',
    Stimmen: '1035107',
    'Wählerzahl': '90035',
    'Wähleranteil %': '29.63',
    'Anzahl Sitze': '54',
    classes: [ 'dunkelblau', 'rechtsbuendig', 'hellblau' ] },
  { 'Listen Nr.': '2',
    'Listen Bez.': 'SP',
    Stimmen: '650843',
    'Wählerzahl': '58687',
    'Wähleranteil %': '19.32',
    'Anzahl Sitze': '35',
    classes: [ 'dunkelblau', 'rechtsbuendig', 'hellblau' ] },
  { 'Listen Nr.': '3',
    'Listen Bez.': 'FDP',
    Stimmen: '447063',
    'Wählerzahl': '39296',
    'Wähleranteil %': '12.93',
    'Anzahl Sitze': '23',
    classes: [ 'dunkelblau', 'rechtsbuendig', 'hellblau' ] },
  { 'Listen Nr.': '4',
    'Listen Bez.': 'GRÜNE',
    Stimmen: '346202',
    'Wählerzahl': '32111',
    'Wähleranteil %': '10.57',
    'Anzahl Sitze': '19',
    classes: [ 'dunkelblau', 'rechtsbuendig', 'hellblau' ] },
  { 'Listen Nr.': '5',
    'Listen Bez.': 'CVP',
    Stimmen: '174451',
    'Wählerzahl': '14774',
    'Wähleranteil %': '4.86',
    'Anzahl Sitze': '9',
    classes: [ 'dunkelblau', 'rechtsbuendig', 'hellblau' ] },
  { 'Listen Nr.': '6',
    'Listen Bez.': 'EVP',
    Stimmen: '129024',
    'Wählerzahl': '11479',
    'Wähleranteil %': '3.78',
    'Anzahl Sitze': '7',
    classes: [ 'dunkelblau', 'rechtsbuendig', 'hellblau' ] },
  { 'Listen Nr.': '7',
    'Listen Bez.': 'glp',
    Stimmen: '354500',
    'Wählerzahl': '31191',
    'Wähleranteil %': '10.27',
    'Anzahl Sitze': '19',
    classes: [ 'dunkelblau', 'rechtsbuendig', 'hellblau' ] },
  { 'Listen Nr.': '8',
    'Listen Bez.': 'EDU',
    Stimmen: '90465',
    'Wählerzahl': '7815',
    'Wähleranteil %': '2.57',
    'Anzahl Sitze': '5',
    classes: [ 'dunkelblau', 'rechtsbuendig', 'hellblau' ] },
  { 'Listen Nr.': '9',
    'Listen Bez.': 'AL',
    Stimmen: '49698',
    'Wählerzahl': '4945',
    'Wähleranteil %': '1.63',
    'Anzahl Sitze': '3',
    classes: [ 'dunkelblau', 'rechtsbuendig', 'hellblau' ] },
  { 'Listen Nr.': '10',
    'Listen Bez.': 'SD',
    Stimmen: '14517',
    'Wählerzahl': '1231',
    'Wähleranteil %': '0.41',
    'Anzahl Sitze': '0',
    classes: [ 'dunkelblau', 'rechtsbuendig', 'hellblau' ] },
  { 'Listen Nr.': '11',
    'Listen Bez.': 'Piraten',
    Stimmen: '20835',
    'Wählerzahl': '1708',
    'Wähleranteil %': '0.56',
    'Anzahl Sitze': '0',
    classes: [ 'dunkelblau', 'rechtsbuendig', 'hellblau' ] },
  { 'Listen Nr.': '12',
    'Listen Bez.': 'ERP',
    Stimmen: '238',
    'Wählerzahl': '15',
    'Wähleranteil %': '0.00',
    'Anzahl Sitze': '0',
    classes: [ 'dunkelblau', 'rechtsbuendig', 'hellblau' ] },
  { 'Listen Nr.': '13',
    'Listen Bez.': 'BDP',
    Stimmen: '125243',
    'Wählerzahl': '10529',
    'Wähleranteil %': '3.47',
    'Anzahl Sitze': '6',
    classes: [ 'dunkelblau', 'rechtsbuendig', 'hellblau' ] } ]
```

**Comment by tpreusse** [comment]
Recommendation

``` javascript
{
  source: {
    time: '2015-04-12T14:15:28Z', // UTC
    area: '39 von 183 Gebieten',
    type: 'provisorisches Zwischenresultat',
    electionId: 'kr2011_medieninfo|rr2015_preview',// do not publish hot ones before April 12th 2015 12:00 PM
    url: 'http://www.wahlen.zh.ch/wahlen/kr2011_medieninfo/viewer.php?menu=...'
  },
  results: [
    // example lists canton
    // source live: http://www.wahlen.zh.ch/wahlen/kr2011_medieninfo/viewer.php?menu=listen_kanton
    // source after: http://www.wahlen.zh.ch/wahlen/kr2011_medieninfo/viewer.php?menu=sitzzuteilung
    {
      id: 'SVP|SP|FDP|Grüne|glp|CVP|EVP|BDP|EDU|AL|IP|ZH|JUSO|Piraten',
      type: 'party',
      level: 'canton',
      levelId: 'ZH',
      quorum: true || false || undefined, // Teilnahme an Sitzzuteilung || Keine Teilnahme an Sitzzuteilung || Zwischenresultat
      seats: 36, // Anzahl Sitze
      votesAbs: 411, // Stimmen
      votesPercent: 19, // Wähleranteil %
      votersCount: 98 // Wählerzahl
    },
    // example lists constituency
    // source live: http://www.wahlen.zh.ch/wahlen/kr2011_medieninfo/viewer.php?menu=listen_wk&wk=a
    // source after: http://www.wahlen.zh.ch/wahlen/kr2011_medieninfo/viewer.php?menu=sitzzuteilung
    {
      id: 'SVP|SP|FDP|Grüne|glp|CVP|EVP|BDP|EDU|AL|IP|ZH|JUSO|Piraten',
      type: 'party',
      level: 'constituency',
      levelId: 'WK1',
      quorum: true || false || undefined, // Quorum erreicht || nicht erreicht || Zwischenresultat
      seats: 2, // Anzahl Sitze
      votesAbs: 411, // Stimmen absolut
      votesPercent: 19, // Stimmen %
      votersCount: 98 // Wählerzahl
    },
    // example RR candidate
    // http://www.wahlen.zh.ch/wahlen/rr2015_preview/viewer.php?table=kandkanton
    // AND http://www.wahlen.zh.ch/wahlen/rr2015_preview/viewer.php?table=kandgemeinden
    // -> needs manual aggregation
    {
      id: 'carmen-walker-spaeh', // first last, transliterate, all lowercase, replace space with dash
      type: 'candidate',
      name: 'Carmen Walker Späh',
      level: 'canton|constituency',
      levelId: 'ZH|WKX',
      party: 'SVP|SP|FDP|Grüne|CVP|EVP|BDP|AL|Unpolitisch|-',
      majority: 100000 || undefined, // Canton: abs. Mehr, Constituency: undefined
      hasMajority: true || false,
        // Canton: abs. Mehr <= Stimmen
        // Constituency: Abs. Mehr im Gebiet erreicht || Abs. Mehr im Gebiet nicht erreicht
      votesAbs: 9998, // Stimmen
      incumbent: true || false, // Bisher
      elected: true || false || undefined
        // Canton: Abs. Mehr erreicht und in den ersten 7 Positionen
        // Constituency: undefined
    },
    // example KR candidate
    // http://www.wahlen.zh.ch/wahlen/kr2011_medieninfo/viewer.php?menu=kand_kanton
    {
      id: '1-101', // Wk + '-' + Nr
      type: 'candidate',
      name: 'Ernst Bachmann', // Vorname + ' ' + Name
      level: 'canton',
      levelId: 'ZH',
      party: 'SVP|SP|FDP|Grüne|glp|CVP|EVP|BDP|EDU|AL|IP|ZH|JUSO|Piraten',
      votesAbs: 9998, // Stimmen
      incumbent: true || false, // Bisher
      elected: true || false // Gew./Abg. === 'gewählt'
    }
  ]
}
```

**Comment by tpreusse** [comment]
I will also want to add historic values, thinking of also crawling:
http://www.wahlen.zh.ch/wahlen/kr2011_medieninfo/viewer.php?menu=sitzzuteilung_vergleich
and
http://www.wahlen.zh.ch/wahlen/kr2011_medieninfo/viewer.php?menu=listen_vergleich_wk&wk=a

would make crawling
http://www.wahlen.zh.ch/wahlen/kr2011_medieninfo/viewer.php?menu=sitzzuteilung
obsolete

**Comment by tpreusse** [comment]
Updated Version:

``` javascript
{
  source: {
    time: '2015-04-12T14:15:28Z', // UTC
    area: '39 von 183 Gebieten',
    type: 'provisorisches Zwischenresultat|definitives Endresultat',
    electionId: 'kr2011_medieninfo|rr2015_preview',// do not publish hot ones before April 12th 2015 12:00 PM
    urls: ['http://www.wahlen.zh.ch/wahlen/kr2011_medieninfo/viewer.php?menu=...']
  },
  results: [
    // example lists canton
    // http://www.wahlen.zh.ch/wahlen/kr2011_medieninfo/viewer.php?menu=listen_kanton
    // AND http://www.wahlen.zh.ch/wahlen/kr2011_medieninfo/viewer.php?menu=listen_vergleich_kanton
    {
      id: 'SVP|SP|FDP|Grüne|glp|CVP|EVP|BDP|EDU|AL|IP|ZH|JUSO|Piraten',
      type: 'party',
      level: 'canton',
      levelId: 'ZH',
      votes: 411, // Stimmen
      voters: 98, // Wählerzahl
      percent: 19, // Wähleranteil %
      quorum: true || false, // Wähleranteil % > 5%
      // listen_vergleich_kanton
      seats: 36, // Sitze 2015
      previousSeats: 35, // Sitze 2011
      previousPercent: 1 // Wähleranteil % 2011
    },
    // example lists constituency
    // http://www.wahlen.zh.ch/wahlen/kr2011_medieninfo/viewer.php?menu=listen_wk&wk=a
    // AND http://www.wahlen.zh.ch/wahlen/kr2011_medieninfo/viewer.php?menu=listen_vergleich_wk&wk=a
    // AND http://www.wahlen.zh.ch/wahlen/kr2011_medieninfo/viewer.php?menu=sitzzuteilung_vergleich
    {
      id: 'SVP|SP|FDP|Grüne|glp|CVP|EVP|BDP|EDU|AL|IP|ZH|JUSO|Piraten',
      type: 'party',
      level: 'constituency',
      levelId: 'WK1',
      votesAbs: 411, // Stimmen absolut
      votersCount: 98, // Wählerzahl
      votesPercent: 19, // Stimmen %
      quorum: true || false, // Stimmen % > 5%
      // sitzzuteilung_vergleich
      seats: 36, // Sitze 2015
      previousSeats: 35, // Sitze 2011
      // listen_vergleich_wk&wk=a
      previousPercent: 1 // % Stimmen 2011
    },
    // example RR candidate
    // http://www.wahlen.zh.ch/wahlen/rr2015_preview/viewer.php?table=kandkanton
    // AND http://www.wahlen.zh.ch/wahlen/rr2015_preview/viewer.php?table=kandgemeinden
    // -> needs manual aggregation
    {
      id: 'carmen-walker-spaeh', // first last, transliterate, all lowercase, replace space with dash
      type: 'candidate',
      name: 'Carmen Walker Späh',
      level: 'canton|constituency',
      levelId: 'ZH|WKX',
      party: 'SVP|SP|FDP|Grüne|CVP|EVP|BDP|AL|Unpolitisch|-',
      majority: 100000 || undefined, // Canton: abs. Mehr, Constituency: undefined
      hasMajority: true || false,
        // Canton: abs. Mehr <= Stimmen
        // Constituency: Abs. Mehr im Gebiet erreicht || Abs. Mehr im Gebiet nicht erreicht
      votesAbs: 9998, // Stimmen
      incumbent: true || false, // Bisher
      elected: true || false || undefined
        // Canton: Abs. Mehr erreicht und in den ersten 7 Positionen
        // Constituency: undefined
    },
    // example KR candidate
    // http://www.wahlen.zh.ch/wahlen/kr2011_medieninfo/viewer.php?menu=kand_kanton
    {
      id: '1-101', // Wk + '-' + Nr
      type: 'candidate',
      name: 'Ernst Bachmann', // Vorname + ' ' + Name
      level: 'canton',
      levelId: 'ZH',
      party: 'SVP|SP|FDP|Grüne|glp|CVP|EVP|BDP|EDU|AL|IP|ZH|JUSO|Piraten',
      votesAbs: 9998, // Stimmen
      incumbent: true || false, // Bisher
      elected: true || false // Gew./Abg. === 'gewählt'
    }
  ]
}
```

**Comment by tpreusse** [comment]
Decided to move `level` information into a geography object:

``` js
{
  id: 'carmen-walker-spaeh',
  type: 'candidate',
  geography: {
    'id': 1, // Aeugst am Albis
    'type': 'area'
  },
  ...
}
```

Types: `canton` (ZH), `constituency` (WK # 1-18), `area` (`BFSK` - BFS `GDENR` + Kreis #).

**Comment by tpreusse** [comment]
Implemented in version 0.0.3, see [README](https://github.com/nzzdev/red-zh-vote-crawler#data-format) for final data format.
